llmcam/fn_to_fc.py

```diff
@@ -139,52 +138,52 @@ def tool_schema(
 
 # %% ../nbs/06_fn_to_fc.ipynb 26
 # Environmental setting up
-tools = [tool_schema(fn) for fn in (capture_youtube_live_frame_and_save, ask_gpt4v_about_image_file)]
-initial_messages = [{"role":"system", "content":"You are a helpful system administrator. Use the supplied tools to assist the user."}]
+YTLiveTools = [tool_schema(fn) for fn in (capture_youtube_live_frame_and_save, ask_gpt4v_about_image_file)]
 
 # %% ../nbs/06_fn_to_fc.ipynb 30
 # Support functions to handle tool response,where res == response.choices[0].message
 def fn_name(res): return res.tool_calls[0].function.name
 def fn_args(res): return json.loads(res.tool_calls[0].function.arguments)    
-def fn_exec(res): return globals().get(fn_name(res))(**fn_args(res))
-def fn_result_content(res):
+def fn_exec(res, aux_fn):
+    fn = globals().get(fn_name(res))
+    if fn: return fn(**fn_args(res))
+    aux_fn(fn_name(res), **fn_args(res))
+    
+def fn_result_content(res, aux_fn):
     """Create a content containing the result of the function call"""
     content = dict()
     content.update(fn_args(res))
-    content.update({fn_name(res): fn_exec(res)})
+    content.update({fn_name(res): fn_exec(res, aux_fn)})
     return json.dumps(content)
 
 # %% ../nbs/06_fn_to_fc.ipynb 31
-def generate_messages(
-    message: str,  # New message frorm the user
-    history : list[dict] = []  # Previous messages
-) -> list[dict]:  # List of messages
-    """Generate messages from the user and the system"""
-    # Copy the history to avoid modifying the original list
-    messages = history.copy()
-    if len(messages) == 0:
-        # Add initial system message if no history
-        messages.append({
-            "role":"system", 
-            "content":"You are a helpful system administrator. Use the supplied tools to assist the user."
-        })
-
-    def complete(
-            role: Literal["system", "user", "tool", "assistant"],  # The role of the message sender
-            content: str,  # The content of the message
-            tool_call_id=None):
-        """Send completion request with messages, and save the response in messages again"""
-        messages.append({"role":role, "content":content, "tool_call_id":tool_call_id})
-        response = openai.chat.completions.create(
-            model="gpt-4o", 
-            messages=messages, 
-            tools=tools
+def complete(
+        messages: list[dict],  # The list of messages
+        role: Literal["system", "user", "assistant", "tool"],  # The role of the message sender
+        content: str,  # The content of the message
```

Comment:
That's because the `role` and `content` are saved separately as hidden text fields for each message (`ChatMessage` class). These fields are submitted along with the new message as fields in a form. I want to keep them separate for now because if saved together they will be of JSON format, which needs to be converted to string in a form. This can cause issue because stringifying JSON can be problematic with special characters or automatic type converter when reading JSON from string.

Reply:
> That's because the `role` and `content` are saved separately as hidden text fields for each message (`ChatMessage` class). These fields are submitted along with the new message as fields in a form. I want to keep them separate for now because if saved together they will be of JSON format, which needs to be converted to string in a form. This can cause issue because stringifying JSON can be problematic with special characters or automatic type converter when reading JSON from string.

Probably cleaner way is use some macro / function to construct json from role & message and append it to messages.